src/main/java/com/mindee/http/MindeeHttpApi.java

```diff
@@ -209,34 +199,52 @@ public <DocT extends Inference> AsyncPredictResponse<DocT> predictAsyncPost(
 
     // required to register jackson date module format to deserialize
     mapper.findAndRegisterModules();
-
-    String errorMessage = "Mindee API client: ";
-    AsyncPredictResponse<DocT> predictResponse;
-
+    JavaType type = mapper.getTypeFactory().constructParametricType(
+        AsyncPredictResponse.class,
+        documentClass
+    );
     try (
         CloseableHttpClient httpClient = httpClientBuilder.build();
         CloseableHttpResponse response = httpClient.execute(post)
     ) {
       HttpEntity responseEntity = response.getEntity();
-      if (responseEntity.getContentLength() != 0) {
-        JavaType type = mapper.getTypeFactory().constructParametricType(
-            AsyncPredictResponse.class,
-            documentClass
-        );
-        predictResponse = mapper.readValue(responseEntity.getContent(), type);
-        if (is2xxStatusCode(response.getStatusLine().getStatusCode())) {
-          return predictResponse;
-        }
-        if (predictResponse != null) {
-          errorMessage += predictResponse.getApiRequest().getError().toString();
-        }
-      } else {
-        throw new MindeeException(parseUnhandledError(responseEntity, response));
+      int statusCode = response.getStatusLine().getStatusCode();
+      if (!is2xxStatusCode(statusCode)) {
+        throw getHttpError(type, response);
       }
+      if (responseEntity.getContentLength() == 0) {
+        throw new MindeeException("Empty response from server.");
+      }
+      return mapper.readValue(responseEntity.getContent(), type);
     } catch (IOException err) {
       throw new MindeeException(err.getMessage(), err);
     }
-    throw new MindeeException(errorMessage);
+  }
+
+  private <ResponseT extends ApiResponse> MindeeHttpException getHttpError(
+      JavaType javaType,
+      CloseableHttpResponse response
+  ) {
+    int statusCode = response.getStatusLine().getStatusCode();
+    String message = "HTTP Status " + statusCode + " - ";
+    String details;
+    String rawResponse;
+    try {
+      rawResponse = readRawResponse(response.getEntity());
+    } catch (IOException err) {
+      message += "Could not read server response, check details.";
+      details = err.getMessage();
+      return new MindeeHttpException(statusCode, message, details);
+    }
+    try {
+      ResponseT predictResponse = mapper.readValue(rawResponse, javaType);
+      message = predictResponse.getApiRequest().getError().getMessage();
+      details = predictResponse.getApiRequest().getError().getDetails().toString();
```

Comment:
Sometimes details doesn't exist on regular errors. Don't think this step should be counted as an "Unhandled server response".

Reply:
done